pkg/csi/cinder/nodeserver.go

```diff
@@ -53,16 +55,22 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	if len(volumeID) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume Volume ID must be provided")
 	}
-	if len(source) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume Staging Target Path must be provided")
-	}
 	if len(targetPath) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume Target Path must be provided")
 	}
 	if volumeCapability == nil {
 		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume Volume Capability must be provided")
 	}
 
+	ephemeralVolume := req.GetVolumeContext()["csi.storage.k8s.io/ephemeral"] == "true"
```

Comment:
one question, how is this key (csi.storage.k8s.io/ephemeral) set , is it set by external-provioner or driver sets based on the mode passed in manifest?

Reply:
if both modes are enabled in a driver(modes can be enabled or disabled by using volumeLifecycleModes in CSIDriver CRD) , then during NodePublishVolume Call we need to know weather it is for a normal volume or inline, it is determined by using this key `csi.storage.k8s.io/ephemeral` which set/unset by kubelet if podInfoMount is true.

Reply:
sorry, Im confused, podInfoMount is always true in driver CRD, and we have both modes enabled , persistent + ephemeral , how kubelet is gonna differentiate when persistent volume is requested , when ephemeral is requested, podInfoMount alone cant give that info right

Reply:
Kubelet is responsible for setting the key `csi.storage.k8s.io/ephemeral`, driver developers can differ between inline persistent volume using this key, how kubelet differs, please read the KEP :)
https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md 
https://kubernetes-csi.github.io/docs/pod-info.html

Reply:
@adisky Ok, got it. Thanks.